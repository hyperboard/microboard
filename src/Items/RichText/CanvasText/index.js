function K(t,n){n||(n=1/0);let e=[];for(let s of t)e.push(m(s,n));P(e);let o=0,i=0;for(let s of e)o=Math.max(o,s.width),i+=s.height;return E(e,n),{nodes:e,maxWidth:n,width:o,height:i,render:s=>{N(s,e)}}}var L={paragraph:m,heading:m,"block-quote":m,"bulleted-list":m,"numbered-list":m,"list-item":m,text:C};function m(t,n){var o;let e={type:t.type,lineHeight:(o=t.lineHeight)!=null?o:1.4,children:[],lines:[],align:t.horisontalAlignment};for(let i of t.children)e.children.push(L[i.type](i));return W(e,n),e}function C(t){let n=t.text.length===0?"\xA0":t.text;return{type:"text",text:n,style:T(t),blocks:[]}}function T(t){var o,i,s;let n={fontStyle:"normal",fontWeight:"normal",color:(o=t.fontColor)!=null?o:"black",backgroundColor:t.fontHighlight,fontSize:(i=t.fontSize)!=null?i:14,fontFamily:(s=t.fontFamily)!=null?s:"Arial"},e=t.styles?t.styles:[];for(let r of e)switch(r){case"bold":n.fontWeight="bold";break;case"italic":n.fontStyle="italic";break;case"underline":n.textDecorationLine="underline";break;case"strikethrough":n.textDecorationLine="line-through";break;case"superscript":n.verticalAlign="super";break;case"subscript":n.verticalAlign="sub";break}return n.font=A(n),n}var B=T({});function A(t){let n=t.fontStyle||"normal",e=t.fontWeight||400,o=t.fontSize||"14",i=t.fontFamily||"Arial";return`${n} ${e} ${o}px ${i}`}function W(t,n){let e=[];for(let o of t.children)F(e,o,n);t.lines=e,z(t),H(t,0,0)}function F(t,n,e){if(n.text===""){return}t.length===0&&(t.push([]),void 0);let o=n.style,i=n.text.split(`
`);for(let s of i){s=s.replaceAll("	","        ");let r=M(s);let l="",u=0,g=!1;for(;r.length>0;){u++;let c=r.shift(),f=l===""?c:l+""+c;let d=x(f,o);let a=t[t.length-1],w=0;for(let y of a)w+=y.width;if(w+d.width<=e)l=f;else if(l==="")if(a.length===0){let y=O(c,o,e),h=c.slice(y.length);let k=x(y,o);a.push(k),t.push([]),g=!0,r.unshift(h)}else t.push([]),g=!0,r.unshift(c);else{a.length===0&&g&&(l=l.trimStart());let y=x(l,o);a.push(y),t.push([]),g=!0,l="",r.unshift(c)}}if(l!==""){let c=x(l,o);t[t.length-1].push(c)}t.push([])}t[t.length-1].length===0&&t.pop()}function z(t){for(let n=0;n<t.lines.length;n++){let e=t.lines[n];if(e.length===0){let o=t.lines[n-1],i=B;o!==void 0&&(i=o[o.length-1].style),e.push(x(" ",i!=null?i:B,0,0))}}}function H(t,n,e){let o=t.lines,i=t.lineHeight,s=0,r=0,l=0,u=0,g=0;for(let c of o){let f=0,d,a=0,w=0,y=0;if(c.length!==0){for(let h of c)h.x=a,(h.fontSize>f||!d)&&(f=h.fontSize,d=h),a+=h.width;g+=f*i,a>u&&(u=a),l=r,r+=f*i,w=f*i-f,s=r-w/2-d.measure.descent;for(let h of c)h.y=s}}t.width=u,t.height=g}function x(t,n,e,o){let i=I(t,n);return{text:t,style:n,width:i.width,x:e,y:o,measure:i,fontSize:parseFloat(n.fontSize)}}var D=document.createElement("canvas"),S=D.getContext("2d"),b={};function R(t){return typeof t=="number"&&isFinite(t)}function p(t,n=0){return R(t)?t:n}function I(t,n){if(b[n.font]&&b[n.font][t])return b[n.font][t];S.font=n.font;let e=S.measureText(t),o=p(e.actualBoundingBoxAscent),i=p(e.actualBoundingBoxDescent),s=p(e.actualBoundingBoxLeft),r=p(e.actualBoundingBoxRight),l=p(e.fontBoundingBoxAscent),u=p(e.fontBoundingBoxDescent),g=p(e.width),c=Math.max(l,o),f=Math.max(u,i),d={actualBoundingBoxAscent:o,actualBoundingBoxDescent:i,actualBoundingBoxLeft:s,actualBoundingBoxRight:r,fontBoundingBoxAscent:l,fontBoundingBoxDescent:u,ascent:c,descent:f,width:Math.max(g,s+r),height:c+f};return b[n.font]||(b[n.font]={}),b[n.font][t]=d,d}function M(t){return t.split(/(\s+)/)}function O(t,n,e){let o=0,i=t.length,s="";for(;o<=i;){let r=Math.floor((o+i)/2);if(r===0){s=t.slice(0,1);break}let l=t.slice(0,r);x(l,n).width<=e?(s=l,o=r+1):i=r-1}return s}function P(t){let n=0;for(let e of t){for(let o of e.lines)for(let i of o)i.y+=n;n+=e.height}}function E(t,n){for(let e of t)switch(e.align){case"left":break;case"center":$(e,n);break;case"right":v(e,n);break}}function $(t,n){let e=t.width;for(let o of t.lines){let i=0;for(let r of o)i+=r.width;let s=(n-i)/2;for(let r of o)r.x+=s}}function v(t,n){let e=t.width;for(let o of t.lines){let i=0;for(let r of o)i+=r.width;let s=n-i;for(let r of o)r.x+=s}}function N(t,n){for(let e of n)U(t,e.lines)}function U(t,n){for(let e of n)for(let o of e)q(t,o)}function q(t,n){t.font=n.style.font,j(t,n),G(t,n),J(t,n)}function j(t,n){if(!n.style.backgroundColor)return;let e=n.measure;t.fillStyle=n.style.backgroundColor,t.fillRect(n.x,n.y-e.ascent,e.width,e.height)}function G(t,n){if(n.style.textDecorationLine!=="underline")return;let e=n.x,o=n.y,i=n.style,s=n.measure,r=s.width,l=s.height,u=i.color;t.strokeStyle=u,t.lineWidth=1,t.beginPath(),t.moveTo(e,o+2),t.lineTo(e+r,o+2),t.stroke(),t.strokeStyle=i.backgroundColor,t.lineWidth=2}function J(t,n){let{text:e,style:o,x:i,y:s,measure:r}=n;t.fillStyle=o.color,t.fillText(e,i,s)}export{K as getBlockNodes};
