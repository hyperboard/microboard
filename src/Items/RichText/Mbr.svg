<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created with Inkscape (http://www.inkscape.org/) -->

<svg
   width="800"
   height="450"
   viewBox="0 0 800 450"
   version="1.1"
   id="SVGRoot"
   sodipodi:docname="Mbr.svg"
   inkscape:version="1.2.1 (9c6d41e410, 2022-07-14, custom)"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:svg="http://www.w3.org/2000/svg">
  <sodipodi:namedview
     id="namedview459"
     pagecolor="#ffffff"
     bordercolor="#666666"
     borderopacity="1.0"
     inkscape:pageshadow="2"
     inkscape:pageopacity="0.0"
     inkscape:pagecheckerboard="0"
     inkscape:document-units="px"
     showgrid="true"
     inkscape:zoom="1.381"
     inkscape:cx="500"
     inkscape:cy="227.73353"
     inkscape:window-width="1920"
     inkscape:window-height="999"
     inkscape:window-x="0"
     inkscape:window-y="44"
     inkscape:window-maximized="1"
     inkscape:current-layer="layer1"
     width="1000px"
     inkscape:showpageshadow="2"
     inkscape:deskcolor="#d1d1d1">
    <inkscape:grid
       type="xygrid"
       id="grid487"
       spacingx="5"
       spacingy="5"
       originx="0"
       originy="0" />
  </sodipodi:namedview>
  <defs
     id="defs454" />
  <g
     inkscape:label="Layer 1"
     inkscape:groupmode="layer"
     id="layer1">
    <text
       xml:space="preserve"
       style="font-size:20px;line-height:1.25;font-family:sans-serif"
       x="49.712902"
       y="48.33432"
       id="text1056"><tspan
         sodipodi:role="line"
         id="tspan1054"
         x="49.712902"
         y="48.33432">HTML</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:20px;line-height:1.25;font-family:sans-serif"
       x="449.47964"
       y="44.82016"
       id="text1060"><tspan
         sodipodi:role="line"
         id="tspan1058"
         x="449.47964"
         y="44.82016">Canvas</tspan></text>
    <g
       id="g2147"
       transform="translate(-74.507352,2.0209223)">
      <g
         id="g1843">
        <rect
           style="fill:none;stroke:#000000;stroke-width:1.03949"
           id="rect1807"
           width="298.51035"
           height="273.25214"
           x="525.73157"
           y="100.60113" />
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="524.37439"
           y="85.237213"
           id="text1811"><tspan
             sodipodi:role="line"
             id="tspan1809"
             x="524.37439"
             y="85.237213">Left, Top</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="744.76654"
           y="400.31729"
           id="text1815"><tspan
             sodipodi:role="line"
             id="tspan1813"
             x="744.76654"
             y="400.31729">Right, Bottom</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="650.30841"
           y="85.176613"
           id="text1819"><tspan
             sodipodi:role="line"
             id="tspan1817"
             x="650.30841"
             y="85.176613">Width</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="-252.66336"
           y="522.79822"
           id="text1823"
           transform="rotate(-88.965426)"><tspan
             sodipodi:role="line"
             id="tspan1821"
             x="-252.66336"
             y="522.79822">Heigth</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="536.8678"
           y="123.72548"
           id="text1827"><tspan
             sodipodi:role="line"
             id="tspan1825"
             x="536.8678"
             y="123.72548">Translate (Left, Top)</tspan><tspan
             sodipodi:role="line"
             x="536.8678"
             y="141.22548"
             id="tspan1829">Scale (Width, Heigth)</tspan></text>
      </g>
      <g
         id="g1843-3"
         transform="matrix(0.59715836,0,0,0.51616836,281.68541,125.2077)">
        <rect
           style="fill:none;stroke:#000000;stroke-width:1.03949"
           id="rect1807-6"
           width="298.51035"
           height="273.25214"
           x="525.73157"
           y="100.60113" />
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="524.37439"
           y="85.237213"
           id="text1811-7"><tspan
             sodipodi:role="line"
             id="tspan1809-5"
             x="524.37439"
             y="85.237213">Left, Top</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="744.76654"
           y="400.31729"
           id="text1815-3"><tspan
             sodipodi:role="line"
             id="tspan1813-5"
             x="744.76654"
             y="400.31729">Right, Bottom</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="650.30841"
           y="85.176613"
           id="text1819-6"><tspan
             sodipodi:role="line"
             id="tspan1817-2"
             x="650.30841"
             y="85.176613">Width</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="-252.66336"
           y="522.79822"
           id="text1823-9"
           transform="rotate(-88.965426)"><tspan
             sodipodi:role="line"
             id="tspan1821-1"
             x="-252.66336"
             y="522.79822">Heigth</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="536.8678"
           y="123.72548"
           id="text1827-2"><tspan
             sodipodi:role="line"
             id="tspan1825-7"
             x="536.8678"
             y="123.72548">Translate (Left, Top)</tspan><tspan
             sodipodi:role="line"
             x="536.8678"
             y="141.22548"
             id="tspan1829-0">Scale (Width, Heigth)</tspan></text>
      </g>
    </g>
    <g
       id="g2147-3"
       transform="translate(-473.90773,-2.116698)">
      <g
         id="g1843-1">
        <rect
           style="fill:none;stroke:#000000;stroke-width:1.03949"
           id="rect1807-9"
           width="298.51035"
           height="273.25214"
           x="525.73157"
           y="100.60113" />
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="524.37439"
           y="85.237213"
           id="text1811-4"><tspan
             sodipodi:role="line"
             id="tspan1809-7"
             x="524.37439"
             y="85.237213">Left, Top</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="744.76654"
           y="400.31729"
           id="text1815-8"><tspan
             sodipodi:role="line"
             id="tspan1813-4"
             x="744.76654"
             y="400.31729">Right, Bottom</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="650.30841"
           y="85.176613"
           id="text1819-5"><tspan
             sodipodi:role="line"
             id="tspan1817-0"
             x="650.30841"
             y="85.176613">Width</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="-252.66336"
           y="522.79822"
           id="text1823-3"
           transform="rotate(-88.965426)"><tspan
             sodipodi:role="line"
             id="tspan1821-6"
             x="-252.66336"
             y="522.79822">Heigth</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="536.8678"
           y="123.72548"
           id="text1827-1"><tspan
             sodipodi:role="line"
             id="tspan1825-0"
             x="536.8678"
             y="123.72548">Translate (Left, Top)</tspan><tspan
             sodipodi:role="line"
             x="536.8678"
             y="141.22548"
             id="tspan1829-6">Scale (Width, Heigth)</tspan></text>
      </g>
      <g
         id="g1843-3-3"
         transform="matrix(0.59715836,0,0,0.51616836,281.68541,125.2077)">
        <rect
           style="fill:none;stroke:#000000;stroke-width:1.03949"
           id="rect1807-6-2"
           width="298.51035"
           height="273.25214"
           x="525.73157"
           y="100.60113" />
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="524.37439"
           y="85.237213"
           id="text1811-7-0"><tspan
             sodipodi:role="line"
             id="tspan1809-5-6"
             x="524.37439"
             y="85.237213">Left, Top</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="744.76654"
           y="400.31729"
           id="text1815-3-1"><tspan
             sodipodi:role="line"
             id="tspan1813-5-5"
             x="744.76654"
             y="400.31729">Right, Bottom</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="650.30841"
           y="85.176613"
           id="text1819-6-5"><tspan
             sodipodi:role="line"
             id="tspan1817-2-4"
             x="650.30841"
             y="85.176613">Width</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="-252.66336"
           y="522.79822"
           id="text1823-9-7"
           transform="rotate(-88.965426)"><tspan
             sodipodi:role="line"
             id="tspan1821-1-6"
             x="-252.66336"
             y="522.79822">Heigth</tspan></text>
        <text
           xml:space="preserve"
           style="font-size:14px;line-height:1.25;font-family:sans-serif"
           x="536.8678"
           y="123.72548"
           id="text1827-2-5"><tspan
             sodipodi:role="line"
             id="tspan1825-7-6"
             x="536.8678"
             y="123.72548">Translate (Left, Top)</tspan><tspan
             sodipodi:role="line"
             x="536.8678"
             y="141.22548"
             id="tspan1829-0-9">Scale (Width, Heigth)</tspan></text>
      </g>
    </g>
  </g>
</svg>
