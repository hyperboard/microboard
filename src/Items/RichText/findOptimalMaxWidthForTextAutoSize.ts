import { getBlockNodes } from "./CanvasText/Render.ts";
import { BlockNode } from "./Editor/BlockNode";

/*
Вот 10 слов из 10 букв (не правда):

1. Автомобиль
2. Библиотека
3. Велосипед
4. Гимнастика
5. Дисциплина
6. Жонглировать
7. Интеллект
8. Концепция
9. Лаборатория
10. Мотивация

Вот 10 слов из 17 букв (не правда):

1. Благотворительность
2. Водонепроницаемость
3. Гастроэнтерология
4. Дезориентированный
5. Землепользователь
6. Интернационализм
7. Кораблестроитель
8. Малоинформативный
9. Невосприимчивость
10. Осведомительница

*/
export function findOptimalMaxWidthForTextAutoSize(
	text: BlockNode[],
	containerWidth: number,
	containerHeight: number,
	initialMaxWidth: number,
	tolerance = 0.05,
): {
	bestMaxWidth: number;
	bestMaxHeight: number;
} {
	const targetRatio = containerWidth / containerHeight;
	let low = 0;
	let high = initialMaxWidth * 3;

	let closestRatioDifference = Infinity;
	let closestWidth = initialMaxWidth;
	let closestHeight = initialMaxWidth / targetRatio;
	const iterations = Math.min(Math.max(3, containerWidth / 80), 15);

	for (let i = 0; i < iterations && low < high; i += 1) {
		const mid = (low + high) / 2;
		const { width: calcWidth, height: calcHeight } = getBlockNodes(
			text,
			mid,
		);

		const currentRatio = calcWidth / calcHeight;
		const ratioDifference = Math.abs(currentRatio - targetRatio);

		if (ratioDifference < closestRatioDifference) {
			closestRatioDifference = ratioDifference;
			closestWidth = calcWidth;
			closestHeight = calcHeight;
		}

		if (ratioDifference <= tolerance) {
			break;
		}

		if (currentRatio < targetRatio) {
			low = mid + 1;
		} else {
			high = mid - 1;
		}
	}

	const scale = Math.min(
		containerWidth / closestWidth,
		containerHeight / closestHeight,
	);
	return {
		bestMaxWidth: initialMaxWidth / scale,
		bestMaxHeight: initialMaxWidth / targetRatio / scale,
	};
}
