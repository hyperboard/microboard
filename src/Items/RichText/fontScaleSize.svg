<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created with Inkscape (http://www.inkscape.org/) -->

<svg
   width="700"
   height="200"
   viewBox="0 0 700 200"
   version="1.1"
   id="SVGRoot"
   sodipodi:docname="fontScaleSize.svg"
   inkscape:version="1.2.1 (9c6d41e410, 2022-07-14, custom)"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:svg="http://www.w3.org/2000/svg">
  <sodipodi:namedview
     id="namedview459"
     pagecolor="#ffffff"
     bordercolor="#666666"
     borderopacity="1.0"
     inkscape:pageshadow="2"
     inkscape:pageopacity="0.0"
     inkscape:pagecheckerboard="0"
     inkscape:document-units="px"
     showgrid="true"
     inkscape:zoom="1.6431953"
     inkscape:cx="377.61792"
     inkscape:cy="131.14692"
     inkscape:window-width="1920"
     inkscape:window-height="999"
     inkscape:window-x="0"
     inkscape:window-y="44"
     inkscape:window-maximized="1"
     inkscape:current-layer="layer1"
     width="1000px"
     inkscape:showpageshadow="2"
     inkscape:deskcolor="#d1d1d1">
    <inkscape:grid
       type="xygrid"
       id="grid487"
       spacingx="5"
       spacingy="5"
       originx="0"
       originy="0" />
  </sodipodi:namedview>
  <defs
     id="defs454" />
  <g
     inkscape:label="Layer 1"
     inkscape:groupmode="layer"
     id="layer1">
    <text
       xml:space="preserve"
       style="font-size:14px;line-height:1.25;font-family:sans-serif"
       x="23.726"
       y="125.206"
       id="text788"><tspan
         sodipodi:role="line"
         id="tspan786"
         x="23.726"
         y="125.206">Paragraph</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:28px;line-height:1.25;font-family:sans-serif"
       x="322.67599"
       y="124.356"
       id="text842"><tspan
         sodipodi:role="line"
         id="tspan840"
         x="322.67599"
         y="124.356">Header</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:28px;line-height:1.25;font-family:sans-serif"
       x="147.452"
       y="125.412"
       id="text788-3"><tspan
         sodipodi:role="line"
         id="tspan786-6"
         x="147.452"
         y="125.412">Paragraph</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:14px;line-height:1.25;font-family:sans-serif"
       x="323.66882"
       y="139.17966"
       id="text788-7"><tspan
         sodipodi:role="line"
         id="tspan786-5"
         x="323.66882"
         y="139.17966">Paragraph</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:56px;line-height:1.25;font-family:sans-serif"
       x="470.35199"
       y="123.712"
       id="text842-3"><tspan
         sodipodi:role="line"
         id="tspan840-5"
         x="470.35199"
         y="123.712">Header</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:28px;line-height:1.25;font-family:sans-serif"
       x="472.68671"
       y="153.27727"
       id="text788-3-6"><tspan
         sodipodi:role="line"
         id="tspan786-6-2"
         x="472.68671"
         y="153.27727">Paragraph</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:14px;line-height:1.25;font-family:sans-serif;fill:#ff0000"
       x="23.754"
       y="74.678001"
       id="text1796"><tspan
         sodipodi:role="line"
         id="tspan1794"
         x="23.754"
         y="74.678001"
         style="font-size:14px">Default font size</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:14px;line-height:1.25;font-family:sans-serif;fill:#ff0000"
       x="148.754"
       y="74.678001"
       id="text1800"><tspan
         sodipodi:role="line"
         id="tspan1798"
         x="148.754"
         y="74.678001"
         style="font-size:14px">Double font size</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:14px;line-height:1.25;font-family:sans-serif;fill:#ff0000;stroke:none"
       x="199.87801"
       y="24.677999"
       id="text1804"><tspan
         sodipodi:role="line"
         id="tspan1802"
         x="199.87801"
         y="24.677999">Default text scale</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:14px;line-height:1.25;font-family:sans-serif;fill:#ff0000"
       x="474.59799"
       y="24.677999"
       id="text1808"><tspan
         sodipodi:role="line"
         id="tspan1806"
         x="474.59799"
         y="24.677999">Double text scale</tspan></text>
    <text
       xml:space="preserve"
       style="font-size:14px;line-height:1.25;font-family:sans-serif;fill:#ff0000"
       x="323.89401"
       y="74.678001"
       id="text1800-3"><tspan
         sodipodi:role="line"
         id="tspan1798-6"
         x="323.89401"
         y="74.678001"
         style="font-size:14px">Both font sizes</tspan></text>
  </g>
</svg>
