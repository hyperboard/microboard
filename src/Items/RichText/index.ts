export { RichText } from "./RichText";
export { RichTextData, RichTextOperation } from "./RichTextOperations";
export { TextStyle } from "./Editor/TextNode";
