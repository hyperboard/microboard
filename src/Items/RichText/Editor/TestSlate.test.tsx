import React from "react";
import { Descendant, Editor, Transforms } from "slate";
import { act } from "react-test-renderer";
import { Slate, Editable } from "slate-react";

import renderer from "react-test-renderer";
import { assert } from "chai";
import { UserService } from "../../../../User";
import { Board } from "../../../Board";
import { RichText } from "../RichText";
import { ProxyHistoryEditor } from "./RichTextEditor";
import { Mbr } from "Board/Items/Mbr";

const createNodeMock = () => ({
	ownerDocument: document,
	getRootNode: () => document,
});

describe("slate-react", () => {
	it("выясняем как слушать группы эвентов из Slate, а не каждый эвент в отдельности;", async () => {
		const server = new ServerMock();
		const firstUserService = new UserService(1, 10);
		const firstBoard = new Board("1", server, firstUserService);
		const secondUserService = new UserService(2, 20);
		const secondBoard = new Board("2", server, secondUserService);

		const firstBoardText = new RichText(
			new Mbr(),
			firstUserService.getNextLocalItemId(),
			firstBoard.events,
		);

		const firstBoardSlateProxy = firstBoardText.editor;
		assert(firstBoardSlateProxy);

		const firstBoardEditor = firstBoardSlateProxy.editorProxy;

		firstBoard.add(firstBoardText);

		renderSlate(firstBoardEditor);

		await act(async () => {
			Transforms.insertNodes(firstBoardEditor, [
				{
					type: "block-quote",
					children: [{ type: "text", text: "first-act: text 1" }],
				},
			]);
		});

		await act(async () => {
			Transforms.insertText(firstBoardEditor, "first-act: text 2");
		});

		await act(async () => {
			Transforms.insertNodes(firstBoardEditor, [
				{
					type: "block-quote",
					children: [{ type: "text", text: "second-act: text 3" }],
				},
			]);
		});

		await act(async () => {
			Transforms.insertText(firstBoardEditor, "second-act: text 4");
		});

		await act(async () => {
			firstBoardEditor.undo();
		});

		await act(async () => {
			firstBoardEditor.redo();
		});

		// 2 renders, one for the main element and one for the split element
		expect(firstBoard.events?.list).toEqual(secondBoard.events?.list);

		const secondBoardText = secondBoard.items.findById(
			firstBoardText.getId(),
		) as RichText;
		assert(secondBoardText);
		const secondBoardSlateProxy = secondBoardText.editor;
		assert(secondBoardSlateProxy);
		expect(firstBoardSlateProxy.editorProxy.operations).toEqual(
			secondBoardSlateProxy.editorProxy.operations,
		);
	});

	it("проверяем выделение текста и применение свойств Slate", async () => {
		const server = new ServerMock();
		const userSerivce = new UserService(1, 10);
		const board = new Board("1", server, userSerivce);

		const richText = new RichText(
			new Mbr(),
			userSerivce.getNextLocalItemId(),
			board.events,
		);
		const slateProxy = richText.editor;
		const editorProxy = slateProxy.editorProxy;

		renderSlate(editorProxy);

		await act(async () => {
			Transforms.insertNodes(editorProxy, [
				{
					type: "block-quote",
					children: [{ type: "text", text: "first-act: text 1" }],
				},
			]);
		});

		await act(async () => {
			Transforms.insertText(editorProxy, "first-act: text 2");
		});

		await act(async () => {
			Transforms.select(editorProxy, [0, 0]);
		});

		assert(editorProxy.marks);
	});
});

function renderSlate(
	editorProxy: ProxyHistoryEditor & Editor,
	// начальные операции Slate
	value: Descendant[] = [],
): void {
	act(() => {
		const root: renderer.ReactTestRenderer = renderer.create(
			<Slate
				editor={editorProxy}
				value={value}
				onChange={result => {
					console.info(result);
				}}
			>
				<Editable />
			</Slate>,
			{ createNodeMock },
		);

		// проверка утверждений
		expect(root.toJSON()).toMatchSnapshot();
	});
}
