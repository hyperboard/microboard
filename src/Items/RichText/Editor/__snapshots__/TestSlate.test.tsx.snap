// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slate-react проверяем выделение текста и применение свойств Slate 1`] = `null`;
